site/bps/BP-22-separate-closing-ledgers-from-opening-ledgers.md

```diff
@@ -0,0 +1,97 @@
+---
+title: "BP-22: Separate closing ledgers from opening ledgers"
+issue: https://github.com/apache/bookkeeper/794
+state: "Under Discussion"
+release: "4.7.0"
+---
+
+### Motivation
+
+In the beginning there was openLedger. Opening a ledger meant also figuring out what the last entry of a ledger should be, and writing it to ZooKeeper. For a long time this was the only way to read a ledger. If a writer was writing to the ledger, then anything it wrote after this point would be lost (fencing was added later). The open operation was the natural place to put recovery, as at this point it was only possible to read a recovered ledger.
+
+openLedgerNoRecovery was added in 2011. This allowed users to read from a ledger as it was being written to, which opened up a bunch of tailing uses cases. Recovery was still the default, because that was what it had always been, and tailing was still considered a secondary usecase. If the user wanted to skip recovery, they'd have to explicitly call the no recovery method.
+
+Then the new API arrived. In the new API, tailing is the primary read use case, and recovery has been demoted to a boolean flag on a builder for the open operation. The user is for the most part unaware of recovery.
+
+However, recovery is still one of the most important aspects of BookKeeper. It is the mechanism on which our Total Order Atomic Broadcast guarantees are built. It deserves to be a bit more prominent than a boolean flag in a builder. It also doesn't help that the terminology is inconsistent. The flag is called withRecovery, while the to check if recovery is needed, we call isClosed. Closed itself is ambiguous because it may refer to the local handle, or it may refer to the state of the ledger.
+
+As tailing is now the primary usecase, we expect that if a writer fails, then whichever node takes over as writer already has a non-recovered ReadHandle open. It would be nice to be able to continue using this Handle to read to the end of the ledger.
```

Comment:
> It would be nice to be able to continue using this Handle to read to the end of the ledger.

I would not say it is "nice". when you are reusing readhandle, you need to deal with a lot of corner cases during transition. it is not "nice" than opening a brand new readhandle.

Reply:
which corner cases? if using the same handle, you seal it, and then read to the end of the handle. Sealing it doesn't actually modify anything in the handle. I don't see where the corner case fits in.

Reply:
I mean if you want to implement this `seal(readhandle)`, you might potentially encounter unknown cases around lac, length, metadata and such. what I don't like this proposal is you are proposing mutating `ReadHandle`. people should just simply use `open` to get a readhandle to tailing, and `open-to-seal` a ledger to get a readhandle to read. The readhandle returned by open calls should be remain as it is, there should be no state transition apply to a read handle. those state transition should happen at applicaiton level, what bookkeeper should provide is just primitives. 

in short, IMO, we should

- not add any method like `seal` or `forceClose` to a ReadHandle
- not add any method like `seal` or `forceClose` to BookKeeper to change a ReadHandle

I mean `CompletableFuture<Void> seal(ReadHandle handle)` is effectively same as `ReadHandle#seal()`.

Reply:
Why are we saying 'tailing is the primary usecase'? We need to have an API OpenForRead  and OpenForReadRecovery. If we want to be deliberate, lets come up with explicit API Names.
OpenforReadRecovery() OpenForReadNoRecovery() / OpenForReadTailing() This way we keep the existing functionality and have non-ambigious API names. Basically there is no OpenForReadOnly()

Reply:
> I mean CompletableFuture<Void> seal(ReadHandle handle) is effectively same as ReadHandle#seal().

No, it's not. In CompletableFuture<Void> seal(ReadHandle handle), the full handle is passed just so that you don't have to do something like, seal(handle.getId(), handle.getPassword(), handle.getDigestType()), _nothing_ in the ReadHandle itself is modified. The ReadHandle _will_ update it's sealed status when it reads from ZK (if it's watching metadata, which in tailing it should be), but this is the same as the ReadHandles on any other node which didn't do the sealing.

I'm thinking a Op + Builder makes more sense now, so we would do something like bk.forceSealedOp().forLedger(handle).execute()
Then we could also have
bk.forceSealedOp().withLedgerId(handle.getId()).withPassword(...).withDigestType(...).execute()
This also lends itself better to log0 usecases (I want the API impact of log0 to be as small as possible), as we could have bk.forceSealedOp().forLedger(handle).creatingNextLedger().execute()

@jvrao This is the impression I got from the #774, that streaming should be the default, hence tailing being primary target usecase.

Reply:
- not we are not saying streaming is the primary purpose. the points in #774 are open by default should be non invasive. 

- I think we shouldn't have a api that would touch ReadHandle. `bk.forceSealedOp.withLedgerId(..)` works for me.

Reply:
> not we are not saying streaming is the primary purpose. the points in #774 are open by default should be non invasive.

This givens precedence to use cases where the ledger is read while it is being written to. (i.e. streaming).

> I think we shouldn't have a api that would touch ReadHandle. bk.forceSealedOp.withLedgerId(..) works for me.

As I said in another comment this morning, it doesn't need ReadHandle, Handle is sufficient. Handle is a conceptually a read-only.
The problem with withLedgerId(..) is that it isn't just withLedgerId(). It's withPassword & withDigestType() as well. So the call would look like:
```
bk.newForceSealedOp().withLedgerId(h.getId()).withPassword(h.getLedgerMetadata().getPassword()).withDigestType(h.getLedgerMetadata().getDigestType()).execute()
```
open-to-seal currently has the same problem.
```
bk.newOpenLedgerOp().withLedgerId(h.getId()).withPassword(h.getLedgerMetadata().getPassword()).withDigestType(h.getLedgerMetadata().getDigestType()).withRecovery(true).execute()
```
That's a lot of text to infer an intent from, and it's easy to miss things.
Compare with:
```
bk.newForceSealedOp().forLedger(h).execute()
```

Reply:
yes. you might write more code. but it is much cleaner. because it enforces people to reopen to get a clean handle after seal. The biggest concern that I have is reusing LedgerHandle after seal. I would prefer : when you moving from a "tailing" readhandle to a "sealed" readhandle, close the "tailing" readhandle, sealit and reopen a branch new one. That is what I have been saying : bookkeeper should only provide primitives, should not involve any state transition. applications use primitives to do their own transition.

```
That's a lot of text to infer an intent from, and it's easy to miss things.
```

well. execute does verification. it prevents you miss things.

Reply:
> it enforces people to reopen to get a clean handle after seal. 

No it doesn't. If the ReadHandle will be watching the metadata, and when it changes will update. Tailing has to be able to read to the end of a sealed ledger, as there may be other nodes, which are not doing the sealing which need to read to the end. The BookKeeper#newForceSealedOp could just as easily be running on a different machine or with a different BookKeeper client. It's an external event happening on the ledger which the ReadHandle points to, and the ReadHandle needs to be able to handle that.

I've no problem with having withLedgerId, withPassword, withDigest, alongside fromLedger(Handle). In fact, that's how I had been thinking of it since I started liking the idea of the builder.

> well. execute does verification. it prevents you miss things.

Execution doesn't verify that you've called withRecovery correctly, or that you've called it at all, which is why I was pushing for defaulting to recovery in the first place. Large, dense blocks of text are hard to verify at a glance, and if something as important as recovery is hidden among it, people are going to make mistakes.